lib/gemstash/S3.rb

```diff
@@ -0,0 +1,247 @@
+      file = gemstash_env.base_file("metadata.yml")
+      unless File.exist?(file)
+        gemstash_env.atomic_write(file) do |f|
+          f.write({ storage_version: Gemstash::S3::VERSION,
+                    gemstash_version: Gemstash::VERSION }.to_yaml)
+        end
+      end
+
+      YAML.load_file(file)
+    end
+
+    private
+
+    def check_storage_version
+      version = Gemstash::S3.metadata[:storage_version]
+      return if version <= Gemstash::S3::VERSION
+
+      raise Gemstash::S3::VersionTooNew.new(@folder, version)
+    end
+
+
+  end
+  class S3Resource
+    include Gemstash::Env::Helper
+    include Gemstash::Logging
+    attr_reader :name, :folder, :client
+    VERSION = 1
+
+    class VersionTooNew < StandardError
+      def initialize(name, folder, version)
+        super("Gemstash resource version #{Gemstash::S3Resource::VERSION} does " \
+               "not support version #{version} for resource #{name.inspect} " \
+               "found at #{folder}")
+      end
+    end
+    def initialize(folder,name,client)
+      @folder = folder
+      @name = name
+      safe_name = sanitize(@name)
+      digest = Digest::MD5.hexdigest(@name)
+      child_folder = "#{safe_name}-#{digest}"
+      @folder = File.join(@folder, child_folder)
+      @client = client
+      @S3resource = Aws::S3::Resource.new(client: @client).bucket(gemstash_env.config[:bucket_name])
+      @properties = nil
+    end
+    def save(content, properties = nil)
+      content.each do |key, value|
+        save_content(key, value)
+      end
+      update_properties(properties)
+      self
+    end
+
+    def exist?(key = nil)
+      if key
+        @S3resource.object(properties_filename).exists? && @S3resource.object(content_filename(key)).exists?
+      else
+        @S3resource.object(properties_filename).exists? && content?
+      end
+    end
+
+    def content(key)
+      @content ||= {}
+      load(key) unless @content.include?(key)
+      @content[key]
+    end
+
+    def properties
+      load_properties
+      @properties || {}
+    end
+
+    def update_properties(props)
+      load_properties(true)
+
+      deep_merge = proc do |_, old_value, new_value|
+        if old_value.is_a?(Hash) && new_value.is_a?(Hash)
+          old_value.merge(new_value, &deep_merge)
+        else
+          new_value
+        end
+      end
+
+      props = properties.merge(props || {}, &deep_merge)
+      save_properties(properties.merge(props || {}))
+      self
+    end
+
+    def property?(*keys)
+      keys.inject(node: properties, result: true) do |memo, key|
+        if memo[:result]
+          memo[:result] = memo[:node].is_a?(Hash) && memo[:node].include?(key)
+          memo[:node] = memo[:node][key] if memo[:result]
+        end
+
+        memo
+      end[:result]
+    end
+
+    def delete(key)
+      return self unless exist?(key)
+
+      begin
+        @S3resource.object(content_filename(key)).delete
+      rescue StandardError => e
```

Comment:
I think we want to catch the specific error that `@S3resource.object(content_filename(key)).delete` can cause so we don't so we can rescue properly

Reply:
Fixed in commit #bfa012a81